src/main/java/frc/robot/subsystems/Elevator.java

```diff
@@ -0,0 +1,206 @@
+  private final ShuffleboardTab m_elevatorTab; 
+  private double m_maxHeight; 
+  private boolean m_enabled; 
+  private boolean m_isCalibrated = false; 
+
+  public Elevator(ShuffleboardTab elevatorTab) {
+    m_elevatorTab = elevatorTab; 
+    m_motor = MotorFactory.createTalonFX(ElevatorConstants.kMotorPort, Constants.kCanivoreCAN);
+    m_motor.setInverted(true);
+    m_motor.setNeutralMode(NeutralMode.Brake);
+    m_talonEncoder = new TalonEncoder(m_motor); 
+    m_talonEncoder.setDistancePerPulse(ElevatorConstants.kDistPerPulse);
+    addChild("motor", m_motor);
+   
+    m_elevatorPID = new PIDController(ElevatorConstants.kP, ElevatorConstants.kI, ElevatorConstants.kD);
+    //m_elevatorTab.add("pid",m_elevatorPID);
+    //m_elevatorTab.addNumber("",()-> getHeight());
+    m_elevatorPID.setTolerance(0.03);
+    m_elevatorFF = new ElevatorFeedforward(ElevatorConstants.kS, ElevatorConstants.kG, ElevatorConstants.kV, ElevatorConstants.kA);
+
+    m_topLimitSwitch = new DigitalInput(ElevatorConstants.kTopLimitSwitchPort); 
+    m_bottomLimitSwitch = new DigitalInput(ElevatorConstants.kBottomLimitSwitchPort); 
+    addChild("Upper Limit",m_topLimitSwitch);
+    addChild("Bottom Limit",m_bottomLimitSwitch);
+
+
+    //TODO: log, addDouble doesn't work. 
+    //LogManager.addDouble("Elevator/error", () -> {return m_pid.getSetpoint() - getElevatorHeight();});
+    
+    //m_motor.setSafetyEnabled(true);
+    setUpElevatorTab();
+    setSetpoint(0.5); 
+  }
+
+  @Override
+  public void periodic() {
+    /**
+     * If we hit the bottom limit switch, then we must be at the zero position.
+     * Thus set the setpoint to 0, then atSetpoint() will be triggered, causing the 
+     * ResetEncoderAtBottom() command to end
+     */
+    if(m_enabled && m_isCalibrated) {
+      double pid = m_elevatorPID.calculate(getHeight());
+      double ff = m_elevatorFF.calculate(ElevatorConstants.kVelocity, ElevatorConstants.kAccel); 
+      ff = 0; 
+      double pidClamped = MathUtil.clamp(pid, -ElevatorConstants.kPowerLimit, ElevatorConstants.kPowerLimit);
+      double finalMotorPower = pidClamped + ff; 
+      
+      set(finalMotorPower);
+    } else {
+      //m_motor.feed();
+    }
+
+  }
+
+  public void close() {
+    // close the ports
+    //TODO: Is there any way to close the ports of the limit switches attached to the motor? 
+    m_topLimitSwitch.close(); 
+    m_bottomLimitSwitch.close(); 
+  }
+
+  public void set(double power) {
+    if(isBottom()){
+      if(power >0 ){ 
+        m_motor.set(power); 
+      }else{
+        m_motor.set(0); 
+      }
+    }
+
+    else if(isTop()){
+      if(power<0){
+        m_motor.set(power); 
+      }else{
+        m_motor.set(0); 
+      }
+    }
+
+    else{
+      m_motor.set(power); 
+    }
+
+  
+  }
+  
+  public void setEnabled(boolean isEnabled){
+    m_enabled = isEnabled; 
+  }
+
+  public void setIsCalibrated(){
+    m_isCalibrated = true; 
+  }
+
+
+  public void setSetpoint(double setpoint){
```

Comment:
not a terrable method name but I think it could be better like "setTagetHeight"

Reply:
code change added